lib/charms/operator_libs_linux/v1/snap.py

```diff
@@ -393,6 +393,20 @@ def connect(
         except CalledProcessError as e:
             raise SnapError("Could not {} for snap [{}]: {}".format(_cmd, self._name, e.stderr))
 
+    def hold(self, duration: Optional[str] = None) -> None:
```

Comment:
I think we should make the signature match the top-level `hold_refresh` signature we just updated, for consistency -- otherwise this one takes the hold time as a string, the other one a number of days. So I recommend we change this to match:

```python
def hold(self, days: int = 90, forever: bool = False):
    ...
```

If later we find we need to specify a more exact hold time, we can update `days` to allow a flexible `str` as well, so users can do `snap.hold('1h')`. Though I kinda doubt we will, so it seems better to match `hold_refresh`'s signature and be consistent.

Reply:
It seems to me that although the UX for snap- vs system-wide holds is not the same, the concept is. For both you can set a hold of a given duration, set a hold forever (indefinitely), or unset the hold. That said, I don't mind a different signature here, as long as it's better and Pythonic. It seems to me `duration: str` is the worst of both worlds: neither better than `hold_refresh` and not Pythonic, and also "stringly typed". You have to know the format `h | m | s` used by snap rather than using a Pythonic value.

So here is a Pythonic-type proposal (untested, but shows the intent):

```python
def hold(duration: Optional[datetime.timedelta] = None):
	"""Add a refresh hold to a snap.

	Args:
		duration: the duration for the hold, or None (the default) to hold
			this snap indefinitely.
	"""
	if duration is None:
		hold_str = "forever"
	else:
		hold_str = ""
		seconds = round(duration.total_seconds())
		if seconds >= 3600:
			hold_str += f"{int(seconds/3600)}h"
			seconds %= 3600
		if seconds >= 60:
			hold_str += f"{int(seconds/60)}m"
			seconds %= 60
		if seconds >= 0:
			hold_str += f"{seconds}s"
	self._snap("refresh", [f"--hold={hold_str}"])
```

Reply:
Alternatively, could avoid the `timedelta` type and do use Pythonic types like this:

```python
def hold(hours: Optional[int] = None, minutes: Optional[int] = None,
         seconds: Optional[int] = None):
	"""Add a refresh hold to a snap.

	If no arguments are provided, hold this snap indefinitely.

	Args:
	    hours: number of hours to hold
	    minutes: number of minutes to hold (added to hours)
	    seconds: number of seconds to hold (added to hours and minutes)
	"""
	if hours is None and minutes is None and seconds is None:
		hold_str = "forever"
	else:
		hold_str = ""
		if hours is not None:
			hold_str += f"{hours}h"
		if minutes is not None:
			hold_str += f"{minutes}m"
		if seconds is not None:
			hold_str += f"{seconds}s"
	self._snap("refresh", [f"--hold={hold_str}"])
```

Reply:
Thanks a lot for the suggestions!

Going for the timedelta, as it can be simplified to be always parsed into seconds. 
Additionally, the second one would have more cases where I would need to check if two of them are being used at the same time or not. If at some point `--hold` can be used with `days, weeks...` we would still have a working method, while second implementation would be becoming worse over time